Automatically calculate number of events in apps

amandalund opened the pull request:
The removes the previously required `max_events` input option from `celer-sim` and `celer-g4`.

amandalund committed:
Automatically calculate number of events in apps

sethrj commented on app/celer-g4/GlobalSetup.cc:
```diff
@@ -151,7 +152,17 @@ void GlobalSetup::ReadInput(std::string const& filename)
+            if (!input_.event_file.empty())
+            {
+                return static_cast<size_type>(
+                    HepMC3PrimaryGenerator(input_.event_file).NumEvents());
```
Each time we open the primary generator it reads the whole file (then reopens it)... is there a reasonable way to keep the generator around?

amandalund replied:
Maybe we can store it in the `GlobalSetup`?

sethrj commented on src/celeritas/io/EventReader.hh:
```diff
@@ -50,23 +50,31 @@ class EventReader : public EventReaderInterface
     result_type operator()() final;
 
+    //! Get total number of events
+    size_type num_events() const { return num_events_; }
```
Since this function is being added to `RootEventReader` too, I think you can go ahead and make it part of the `EventReaderInterface`?

amandalund committed:
Address review feedback

amandalund commented:
@sethrj do you know what I'm missing?
```
/home/runner/work/celeritas/celeritas/src/celeritas/io/EventReader.cc:46:26: error: no member named 'skip' in 'HepMC3::Reader'
            temp_reader->skip(1);
```

sethrj commented:
@amandalund I think the problem is that the HepMC3 available from debian (looks like [3.1.2](https://packages.ubuntu.com/focal/libhepmc3-dev), but I'm not sure why the cmake config isn't echoing the version in the CI) is old, and previously only the `accel` library used the `skip` function (which looks like it was first released in [3.2.0]), and it isn't compiled because the "fast" VM doesn't have Geant4. So I guess we could:
1. Require HepMC3 3.2.0 and newer, and disable it on the CI (not great because then we can't run the `app/` unit tests)
2. Add include guards using `#define HEPMC3_VERSION_CODE 3001002` and replace skip with `readevent` where needed (?)
3. Install a newer hepmc3 as part of the CI?

amandalund commented:
Ahh, thank you @sethrj! Maybe I'll add the include guards for now, and if we want to require a newer HepMC3 version we can revert that later.

amandalund committed:
Fix counting of events from HepMC3 file

- `Reader::skip()` was added in HepMC3 version 3.2.0: use `read_event()` to
  count with older versions
- Fix count and validate for zero events

amandalund committed:
Missing include

sethrj reviewed (approved):
Looks great. Thanks @amandalund !
